plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -158,19 +158,19 @@ type CekValEnv uni fun = UniqueMap TermUnique (CekValue uni fun)
 -- defers to the function stored in the environment). This makes the budgeting machinery extensible
 -- and allows us to separate budgeting logic from evaluation logic and avoid branching on the union
 -- of all possible budgeting state types during evaluation.
-newtype CekBudgetSpender cost uni fun = CekBudgetSpender
+newtype CekBudgetSpender cost uni fun s = CekBudgetSpender
```

Comment:
I know you said not to look, but... why not:

```
unCekBudgetSpender :: forall term s . ToExMemory term => STRef ExBudget -> ExBudgetCategory fun -> ExBudget -> CekCarryingM cost term uni fun s()
```
i.e. your spender needs to get passed the STRef when you use it, but that seems fine?

Reply:
The point being, you avoid all the `forall s . ExBudgetMode .. s` stuff, and it's a bit more obvious that in the main implementation we're just passing a `STRef` around and feeding it into this.

Also it'd be easy to get out the final state: just look into the ref!

Reply:
> Also it'd be easy to get out the final state: just look into the ref!

It's designed to be extended to work with two refs (restricting) as well as a single one (counting, tallying).

Reply:
Ah right, I see the issue. I guess I was thinking about what Ed was suggesting with GHC unboxing `STRef`s into mutable variables, which might be inhibited by this kind of abstraction, versus just passing `STRef`s as function arguments.

Reply:
> Ah right, I see the issue. I guess I was thinking about what Ed was suggesting with GHC unboxing STRefs into mutable variables, which might be inhibited by this kind of abstraction, versus just passing STRefs as function arguments.

I don't really know how a mutable variable looks like... Do you happen to have a reference? In my solution all the `STRef`s are within a single definition, so it's kinda like encapsulation which might help optimizations. Or might not indeed.

Reply:
```
data STRef s a = STRef (MutVar# s a)
```
`MutVar` is the actual primitive type. So I guess if you want to get all the way down to `MutVar s Int` or whatever, which is presumably about as performant as it's possible to get in Haskell, you need to persuade Haskell to unbox the `STRef`, which means probably worker-wrapper, which means having them hanging around as function args, I think.

Reply:
But I'm guessing based on some off-hand remarks, we can also worry about it later.

Reply:
Ah, I see. Well, I'd expect that optimization to fire more reliably when the `STRef` is not visible to the outside at all and indeed here's what I see when I compile a version with two `STRef`s:

```
-- RHS size: {terms: 110, types: 236, coercions: 85, joins: 0/1}
UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.$wrestricting
  :: forall s (uni :: * -> *) fun.
     ExCPU
     -> ExMemory
     -> GHC.Prim.State# s
     -> (# GHC.Prim.State# s, ExBudgetMode RestrictingSt uni fun s #)
UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.$wrestricting
  = \ (@ s)
      (@ (uni :: * -> *))
      (@ fun)
      (ww
         :: ExCPU
         Unf=OtherCon [])
      (ww1
         :: ExMemory
         Unf=OtherCon [])
      (w :: GHC.Prim.State# s) ->
      case GHC.Prim.newMutVar# ww w of { (# ipv, ipv1 #) ->
      case GHC.Prim.newMutVar# ww1 ipv of { (# ipv2, ipv3 #) ->
      (# ipv2,
         UntypedPlutusCore.Evaluation.Machine.Cek.Internal.ExBudgetMode
           ((\ (@ term)
               _
               _
               (eta1 :: ExBudget)
               _
               (eta3 :: GHC.Prim.State# s) ->
               case eta1 of { ExBudget cpuToSpend memToSpend ->
               case GHC.Prim.readMutVar# ipv1 eta3 of { (# ipv4, ipv5 #) ->
               case GHC.Prim.readMutVar# ipv3 ipv4 of { (# ipv6, ipv7 #) ->
               let {
                 nt :: Integer
                 nt
                   = integer-gmp-1.0.3.0:GHC.Integer.Type.minusInteger
                       (ipv5 `cast` <Co:1>) (cpuToSpend `cast` <Co:1>) } in
               case GHC.Prim.writeMutVar# ipv1 (nt `cast` <Co:2>) ipv6 of s2#
               { __DEFAULT ->
               case integer-gmp-1.0.3.0:GHC.Integer.Type.minusInteger
                      (ipv7 `cast` <Co:1>) (memToSpend `cast` <Co:1>)
               of nt1
               { __DEFAULT ->
               case GHC.Prim.writeMutVar# ipv3 (nt1 `cast` <Co:2>) s2# of s2#1
               { __DEFAULT ->
               case integer-gmp-1.0.3.0:GHC.Integer.Type.ltInteger#
                      nt
                      UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.restricting2
               of {
                 __DEFAULT ->
                   case integer-gmp-1.0.3.0:GHC.Integer.Type.ltInteger#
                          nt1
                          UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.restricting2
                   of {
                     __DEFAULT ->
                       (# s2#1,
                          UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.counting2 #);
                     1# ->
                       (# s2#1,
                          Data.Either.Left
                            (PlutusCore.Evaluation.Machine.Exception.ErrorWithCause
                               (PlutusCore.Evaluation.Machine.Exception.UserEvaluationError
                                  (UntypedPlutusCore.Evaluation.Machine.Cek.Internal.CekOutOfExError
                                     ((PlutusCore.Evaluation.Machine.ExBudget.ExBudget
                                         (nt `cast` <Co:2>) (nt1 `cast` <Co:2>))
                                      `cast` <Co:2>)))
                               GHC.Maybe.Nothing) #)
                   };
                 1# ->
                   (# s2#1,
                      Data.Either.Left
                        (PlutusCore.Evaluation.Machine.Exception.ErrorWithCause
                           (PlutusCore.Evaluation.Machine.Exception.UserEvaluationError
                              (UntypedPlutusCore.Evaluation.Machine.Cek.Internal.CekOutOfExError
                                 ((PlutusCore.Evaluation.Machine.ExBudget.ExBudget
                                     (nt `cast` <Co:2>) (nt1 `cast` <Co:2>))
                                  `cast` <Co:2>)))
                           GHC.Maybe.Nothing) #)
               }
               }
               }
               }
               }
               }
               })
            `cast` <Co:56>)
           ((\ (s1 :: GHC.Prim.State# s) ->
               case GHC.Prim.readMutVar# ipv1 s1 of { (# ipv4, ipv5 #) ->
               case GHC.Prim.readMutVar# ipv3 ipv4 of { (# ipv6, ipv7 #) ->
               (# ipv6,
                  (PlutusCore.Evaluation.Machine.ExBudget.$WExBudget ipv5 ipv7)
                  `cast` <Co:5> #)
               }
               })
            `cast` <Co:4>) #)
      }
      }
```

`STRef`s do not even get created, there's not a single mention of `STRef` in the file.

Reply:
Nice!